src/rust/engine/fs/src/store.rs

```diff
@@ -74,24 +112,107 @@ impl Store {
-          "LMDB corruption: Directory bytes were not valid",
-        );
-        directory
-      }),
+        directory.merge_from_bytes(bytes).map_err(|e| {
+          format!(
+            "LMDB corruption: Directory bytes for {} were not valid: {:?}",
+            fingerprint_copy,
+            e
+          )
+        })?;
+        Ok(directory)
+      },
+      move |bytes: &[u8]| {
+        let mut directory = bazel_protos::remote_execution::Directory::new();
+        directory.merge_from_bytes(bytes).map_err(|e| {
+          format!(
+            "CAS returned Directory proto for {} which was not valid: {:?}",
+            fingerprint_copy2,
+            e
+          )
+        })?;
+        bazel_protos::verify_directory_canonical(&directory)?;
+        Ok(directory)
+      },
     )
   }
+
+  fn load_bytes_with<
+    T: Send + 'static,
+    FLocal: Fn(&[u8]) -> Result<T, String> + Send + Sync + 'static,
+    FRemote: Fn(&[u8]) -> Result<T, String> + Send + Sync + 'static,
+  >(
+    &self,
+    entry_type: EntryType,
+    fingerprint: Fingerprint,
+    f_local: FLocal,
+    f_remote: FRemote,
+  ) -> BoxFuture<Option<T>, String> {
+    let local = self.local.clone();
+    let maybe_remote = self.remote.clone();
+    self
+      .local
+      .load_bytes_with(entry_type.clone(), fingerprint.clone(), f_local)
+      .and_then(move |maybe_local_value| match maybe_local_value {
+        Some(value_result) => {
+          future::done(value_result.map(|v| Some(v))).to_boxed() as BoxFuture<_, _>
+        }
+        None => {
+          match maybe_remote {
+            Some(remote) => {
+              remote
+                .load_bytes_with(entry_type.clone(), fingerprint, move |bytes: &[u8]| {
+                  Vec::from(bytes)
+                })
+                .and_then(move |maybe_bytes: Option<Vec<u8>>| match maybe_bytes {
+                  Some(bytes) => {
+                    future::done(f_remote(&bytes))
+                      .and_then(move |value| {
+                        local.store_bytes(entry_type, bytes).and_then(
+                          move |digest| if digest.0 ==
+                            fingerprint
+                          {
+                            Ok(Some(value))
+                          } else {
+                            Err(format!(
+                              "CAS gave wrong fingerprint: expected {}, got {}",
+                              fingerprint,
+                              digest.0
+                            ))
+                          },
+                        )
+                      })
+                      .to_boxed()
+                  }
+                  None => future::ok(None).to_boxed() as BoxFuture<_, _>,
+                })
+                .to_boxed()
+            }
+            None => future::ok(None).to_boxed() as BoxFuture<_, _>,
+          }
+        }
+      })
+      .to_boxed()
+  }
 }
 
+// Only public for testing.
```

Comment:
Because this type is so tiny (smaller than a pointer certainly), I think you should probably just mark it `Copy` instead, and remove `Clone`. That will allow it to be implicitly copied.